File: Scene/SceneConstructor.vb
Imports Data
Imports Data.Classes
Imports Entities
Imports OrbitalMechanics.CelestialObjects
Imports OrbitalMechanics.Classes
Imports TurnTracker

Public Class SceneConstructor

    Private ReadOnly _entityManager As IEntityManager
    Private ReadOnly _turnTracker As ITurnTracker
    Private ReadOnly _dataProvider As IDataProvider


    Public Sub New(entityManager As IEntityManager,
                   turnTracker As ITurnTracker,
                   dataProvider As IDataProvider)
        _entityManager = entityManager
        _turnTracker = turnTracker
        _dataProvider = dataProvider
    End Sub


    Private Function RecursiveCreateCelestialObject(o As CelestialObjectData) As ICelestialObject
        Dim primary As BaseCelestialObject = CreateCorrectCelestialObject(o)

        If o.Satellites.Any() Then
            For Each satellite In o.Satellites
                Dim s = RecursiveCreateCelestialObject(satellite)
                If s IsNot Nothing Then
                    Dim orbit = New Orbit(_turnTracker, satellite.Orbit, True)
                    primary.AddSatellite(DirectCast(s, OrbitingCelestialObjectBase), orbit)
                End If
            Next
        End If

        Return primary
    End Function

    Private Function CreateCorrectCelestialObject(o As CelestialObjectData) As BaseCelestialObject
        Select Case o.Physical.Type
            Case CelestialObjectType.Star
                Return New Star(5500, o.Physical, _entityManager)
            Case CelestialObjectType.Planet
                Return New Planet(o.Physical, _entityManager)
            Case CelestialObjectType.Moon
                Return New Moon(o.Physical, _entityManager)
            Case Else
                Return Nothing
        End Select
    End Function

    Public Function SolSystem() As List(Of ICelestialObject)
        Dim system = _dataProvider.SolarSystem
        Return (From o In system Select RecursiveCreateCelestialObject(o)).ToList()
    End Function

End Class

File: space-game/Scene/SceneConstructor.cs
using Data;
using Data.Classes;
using Entities;
using OrbitalMechanics.CelestialObjects;
using OrbitalMechanics.Classes;
using System.Collections.Generic;
using System.Linq;
using TurnTracker;

namespace Scene
{
    public class SceneConstructor
    {
        private readonly IEntityManager _entityManager;
        private readonly ITurnTracker _turnTracker;
        private readonly IDataProvider _dataProvider;

        public SceneConstructor(IEntityManager entityManager, ITurnTracker turnTracker, IDataProvider dataProvider)
        {
            _entityManager = entityManager;
            _turnTracker = turnTracker;
            _dataProvider = dataProvider;
        }

        private ICelestialObject RecursiveCreateCelestialObject(CelestialObjectData o)
        {
            BaseCelestialObject primary = CreateCorrectCelestialObject(o);

            if (o.Satellites.Any())
            {
                foreach (CelestialObjectData satellite in o.Satellites)
                {
                    ICelestialObject s = RecursiveCreateCelestialObject(satellite);
                    if (s != null)
                    {
                        Orbit orbit = new Orbit(_turnTracker, satellite.Orbit, true);
                        primary.AddSatellite((OrbitingCelestialObjectBase)s, orbit);
                    }
                }
            }
            return primary;
        }

        private BaseCelestialObject CreateCorrectCelestialObject(CelestialObjectData o)
        {
            switch (o.Physical.Type)
            {
                case CelestialObjectType.Star:
                    return new Star(5500, o.Physical, _entityManager);
                case CelestialObjectType.Planet:
                    return new Planet(o.Physical, _entityManager);
                case CelestialObjectType.Moon:
                    return new Moon(o.Physical, _entityManager);
                default:
                    return null;
            }
        }

        public List<ICelestialObject> SolSystem()
        {
            List<CelestialObjectData> system = _dataProvider.SolarSystem;
            return (from o in system select RecursiveCreateCelestialObject(o)).ToList();
        }
    }
}
